nengo/spa/actions.py

```diff
@@ -1,151 +1,162 @@
-        "0.5*(dot(vision, DOG) + dot(memory, CAT*MOUSE)*3 - 1)".
+    vocabs : dict
+        Vocabularies to make accessible in the parsed action rules.
     """
 
-    def __init__(self, sources, expression):
-        self.objects = {}   # the list of known terms
-
-        # make all the module outputs as known terms
-        for name in sources:
-            self.objects[name] = Source(name)
-        # handle the term 'dot(a, b)' to mean DotProduct(a, b)
-        self.objects['dot'] = DotProduct
-
-        # use Python's eval to do the parsing of expressions for us
-        self.validate_string(expression)
-        sanitized_exp = ' '.join(expression.split('\n'))
+    builtins = {
+        'dot': DotProduct,
+        'reinterpret': Reinterpret,
+        'translate': Translate}
+
+    def __init__(self, vocabs=None):
+        if vocabs is None:
+            vocabs = {}
+        self.vocabs = vocabs
+
+    def parse_action(self, action, index=0, name=None, strict=True):
+        """Parse an SPA action.
+
+        Parameters
+        ----------
+        action : str
+            Action to parse.
+        index : int, optional
+            Index of the action for identification by basal ganglia and
+            thalamus.
+        name : str, optional
+            Name of the action.
+        strict : bool, optional
+            If ``True`` only actual conditional actions are allowed and an
+            exception will be raised for anything else. If ``False``, allows
+            also the parsing of effects without the conditional part.
+
+        Returns
+        -------
+        :class:`spa_ast.Action` or :class:`spa_ast.Effects`
+        """
         try:
-            self.expression = eval(sanitized_exp, {}, self)
-        except NameError as e:
-            raise SpaParseError("Unknown module in expression '%s': %s" %
-                                (expression, e))
-        except TypeError as e:
-            raise SpaParseError("Invalid operator in expression '%s': %s" %
-                                (expression, e))
-
-        # normalize the result to a summation
-        if not isinstance(self.expression, Summation):
-            self.expression = Summation([self.expression])
-
-    def validate_string(self, text):
-        m = re.search('~[^a-zA-Z]', text)
-        if m is not None:
-            raise SpaParseError("~ is only permitted before names (e.g., DOG) "
-                                "or modules (e.g., vision): %s" % text)
+            condition, effects = action.split('-->', 1)
+        except ValueError:
+            if strict:
+                raise SpaParseError("Not an action, '-->' missing.")
+            else:
+                return self.parse_effects(action, channeled=False)
+        else:
+            return Action(
+                self.parse_expr(condition),
+                self.parse_effects(effects, channeled=True), index)
+
+    def parse_effects(self, effects, channeled=False):
```

Comment:
Using a more generic parser/lexer that uses a lookup table in a PDA/DFA respectively would prevent all this branching. But I'm not sure if this would be useful for the main repo, and we seem morally opposed to importing additional dependencies. For now you can just add `noqa: C901`?